File: src/main/java/com/cj/messagestreaming/Kafka.java
package com.cj.messagestreaming;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import kafka.consumer.ConsumerIterator;
import kafka.consumer.KafkaStream;
import kafka.javaapi.consumer.ConsumerConnector;
import kafka.message.MessageAndMetadata;
import scala.collection.JavaConverters;



public class Kafka {
	
    public static Types.Subscription subscribe(KafkaConfig config) {
        return (id) -> subscribe(config, id);

    }


	public static Types.Publication publish(KafkaConfig config) {
	    return (data) -> {};
    }
	
	

    private static Stream<byte[]> subscribe(KafkaConfig config, String id) {
        final Integer numThreads = 1;
        final String topic = "myCoolTopic";
        ConsumerConnector consumer = kafka.consumer.Consumer.createJavaConsumerConnector(config.getConsumerConfig());
        Map<String, List<KafkaStream<byte[], byte[]>>> consumerMap = consumer.createMessageStreams(new HashMap<String, Integer>() {{
            put(topic, numThreads);
        }});
        final List<KafkaStream<byte[], byte[]>> streams = consumerMap.get(topic);

        final ConsumerIterator<byte[], byte[]> it = streams.get(0).iterator();


        Runtime.getRuntime().addShutdownHook(new Thread(){
            @Override
            public void run() {
                consumer.shutdown();
            }
        });
        
        Iterable<MessageAndMetadata<byte[], byte[]>> i = JavaConverters.asJavaIterableConverter(it.toIterable()).asJava();
    	return StreamSupport.stream(i.spliterator(), false).map(metadata->metadata.message());
	}

}

File: src/main/scala/com/cj/messagestreaming/kafka/Kafka.scala
package com.cj.messagestreaming.kafka

import com.cj.messagestreaming.{Publication, Subscription}
import kafka.consumer.{ConsumerConfig, ConsumerConnector, ConsumerIterator, KafkaStream}
import kafka.message.MessageAndMetadata
import kafka.server.KafkaConfig

import scala.collection.Map

object Kafka {

  type KafkaStreamOfBytes = KafkaStream[Array[Byte], Array[Byte]]

  def makeSubscription(config: ConsumerConfig, topic: String): Subscription = {
    val numThreads: Integer = 1
    val consumer: ConsumerConnector = kafka.consumer.Consumer.create(config)

    val consumerMap: Map[String, List[KafkaStreamOfBytes]] = consumer.createMessageStreams(Map((topic, numThreads)))

    Runtime.getRuntime.addShutdownHook(new Thread() {
      override def run() {
        consumer.shutdown()
      }
    })

//    val streams: List[KafkaStreamOfBytes] = consumerMap.getOrElse(topic, List())
//
//    streams match {
//      case head :: _ => head.iterator.toStream.map(_.message)
//      case _ => Stream()
//    }

    def getMessage(ksb:List[KafkaStreamOfBytes]) = ksb.headOption.map(_.map(_.message))
    consumerMap.get(topic).flatMap(getMessage).getOrElse(Stream()).toStream

    (for {
      x <- consumerMap.get(topic)
      y <- x.headOption
    } yield y.map(_.message).toStream).getOrElse(Stream())

  }

  def makePublication(config: KafkaConfig): Publication = {
    ???
  }
}
